src/main/java/org/broadinstitute/hellbender/tools/copynumber/PreprocessIntervals.java

```diff
@@ -184,30 +233,109 @@ private static IntervalList padIntervals(final List<SimpleInterval> inputInterva
         return paddedIntervalList;
     }
 
-    private static IntervalList generateBins(final IntervalList preparedIntervalList, final int binLength, final SAMSequenceDictionary sequenceDictionary) {
+    private static Stream<Interval> generateBinStream(final IntervalList preparedIntervalList, final int binLength, final boolean grid,
+                                             final int minimumBinLength) {
         if (binLength == 0) {
-            return IntervalList.copyOf(preparedIntervalList);
+            return Utils.stream(preparedIntervalList);
         }
-        final IntervalList bins = new IntervalList(sequenceDictionary);
-        for (final Interval interval : preparedIntervalList) {
-            for (int binStart = interval.getStart(); binStart <= interval.getEnd(); binStart += binLength) {
-                final int binEnd = FastMath.min(binStart + binLength - 1, interval.getEnd());
-                bins.add(new Interval(interval.getContig(), binStart, binEnd));
-            }
-        }
-        return bins;
+        return Utils.stream(preparedIntervalList)
+                .flatMap(interval ->
+                        StreamSupport.stream(
+                                new IntervalSpliterator(interval, binLength, minimumBinLength, grid), false));
     }
 
-    private static IntervalList filterBinsContainingOnlyNs(final IntervalList unfilteredBins, final ReferenceDataSource reference) {
-        final IntervalList bins = new IntervalList(reference.getSequenceDictionary());
-        for (final Interval unfilteredBin : unfilteredBins) {
-            if (!Utils.stream(reference.query(new SimpleInterval(unfilteredBin))).allMatch(b -> Nucleotide.decode(b) == Nucleotide.N)) {
-                bins.add(unfilteredBin);
+    private static Stream<Interval> filterBinsContainingOnlyNs(final Stream<Interval> unfilteredBins, final ReferenceDataSource reference) {
+        return unfilteredBins.filter(interval -> {
+            final byte[] bases = reference.queryAndPrefetch(interval.getContig(), interval.getStart(), interval.getEnd()).getBases();
+            for (byte b : bases) {
+                if (Nucleotide.decode(b) != Nucleotide.N) {
+                    return true;
+                }
             }
-        }
-        return bins;
+            return false;
+        });
     }
 
     @Override
     public void traverse() {}  // no traversal for this tool
+
+    private static class IntervalSpliterator implements Spliterator<Interval> {
```

Comment:
Note also IntervalUtils.cutToShards (which was used in SparkGenomeReadCounts) and related methods.  Can you check to see if an existing method already suffices?